### PR TITLE
Character-3: added find-character-form, additional changes according to the previous PR

### PR DESCRIPTION
**Pull request by Ellyia**

### CODE REVIEW

File: src/app/core/components/spinner/spinner.component.html [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<div *ngIf="this.spinnerService.getLoader()" class="cssload-container">
```
Comment (Phantomus) [inline src/app/core/components/spinner/spinner.component.html]: it is a bad practice to call the service methods directly in the template

File: src/app/core/services/spinner.service.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+@Injectable({
+  providedIn: 'root'
+})
+export class SpinnerService {
+  private loading: boolean = false;
+
+  constructor() { }
+
+  setLoader(loading: boolean) {
```
Comment (Phantomus) [inline src/app/core/services/spinner.service.ts]: try to rewrite it with getter and setter

File: src/app/features/characters/characters.component.html [hunk 3]
```diff
@@ -2,20 +2,27 @@
...
-  <div div class="wrapper-all-chars">
-    <div class="wrapper-characters">
-      <app-character-card
-        *ngFor="let char of list$ | async"
-        [char]="char"
-        (onCardClick)="showCharacter($event)"
+  <div class="wrapper-all-chars">
+    <div>
+      <div class="wrapper-characters"
+      infiniteScroll
```
Comment (Phantomus) [inline src/app/features/characters/characters.component.html]: need to fix attributes tabulation

File: src/app/features/characters/characters.component.ts [hunk 4]
```diff
@@ -1,37 +1,65 @@
-import { CommonModule } from '@angular/common';
+import { NgFor, NgIf } from '@angular/common';
```
Comment (Phantomus) [inline src/app/features/characters/characters.component.ts]: btw, you do not need to import these at all if you will use new @if and @for in the template)

File: src/app/features/characters/characters.component.ts [hunk 5]
```diff
@@ -1,37 +1,65 @@
...
+    CharacterInfoComponent,
+    InfiniteScrollModule,
+    FindCharacterFormComponent
+  ],
   templateUrl: './characters.component.html',
   styleUrl: './characters.component.scss'
 })
 export class CharactersComponent implements OnDestroy {
-  list$!: Observable<MarvelChar[]>;
+  list$: CustomMarvelChar[] = [];
```
Comment (Phantomus) [inline src/app/features/characters/characters.component.ts]: currently, this variable is not Observable

File: src/app/features/characters/characters.component.ts [hunk 6]
```diff
@@ -1,37 +1,65 @@
...
 
   constructor(private charactersService: CharactersMarvelService){
-    this.list$ = this.charactersService.getCharacters();
+    this.showCharacters();
+  }
+
+  showCharacters(): void {
+    this.subs = this.charactersService.getCharacters(this.charsLimit, this.charsOffset)
+      .subscribe(
+        resp => {
```
Comment (Phantomus) [inline src/app/features/characters/characters.component.ts]: please do not split `.subscribe(` and its params on different lines

File: src/app/features/characters/components/character-info/character-info.component.html [hunk 7]
```diff
@@ -1,21 +1,22 @@
...
     </div>
   </div>
 
   <p>{{character?.description}}</p>
 
-  <ul> <span>COMICS:</span>
-    <li *ngFor="let comic of character?.comics?.items">{{comic?.name}}</li>
+  <ul>
+    <h5>Comics:</h5>
+    <li *ngFor="let comic of character?.comicsItems">{{comic?.name}}</li>
```
Comment (Phantomus) [inline src/app/features/characters/components/character-info/character-info.component.html]: try to always write `trackBy` in ngFor, this is good for performance. Also in the new @for syntax `track` is required

File: src/app/features/characters/components/find-character-form/find-character-form.component.html [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    <button (click)="searchCharacter()">FIND</button>
+  </form>
+
+  @switch (msg) {
+    @case (enumMessages.found) {
+      <div class="visit">
+        <p>I've found {{character.name}} page. You can visit it... =></p>
+        <button (click)="navigateToCharPage(character)">{{character.name}} page</button>
+      </div>}
+    @case (enumMessages.emptyInput) {<p>Tis field is required</p>}
```
Comment (Phantomus) [inline src/app/features/characters/components/find-character-form/find-character-form.component.html]: typo in This
Also for better reading, it is good to add extra space between { and <
like `{ <p>Tis field is required</p> }`

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
+import { Component, OnDestroy } from '@angular/core';
+import { FormControl, FormGroup, ReactiveFormsModule } from '@angular/forms';
+import { Router } from '@angular/router';
+import { MarvelChar } from '@characters/models/characters.model';
+import { CharactersMarvelService } from '@characters/services/characters-marvel.service';
+import { Subscription } from 'rxjs';
+
+enum EMSG {
```
Comment (Phantomus) [inline src/app/features/characters/components/find-character-form/find-character-form.component.ts]: 1. It is better to move the enum to the models folder and give it a more descriptive name.
2. Also according to best practices, enum keywords should start in upper case.
3. If you do not receive this string values from the API and you do not use this messages on UI it is better to use numeric emun here

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  selector: 'find-character-form',
+  standalone: true,
+  imports: [ ReactiveFormsModule ],
+  templateUrl: './find-character-form.component.html',
+  styleUrl: './find-character-form.component.scss'
+})
+export class FindCharacterFormComponent implements OnDestroy {
+  static EMSG = EMSG;
+
+  get enumMessages() {
```
Comment (Phantomus) [inline src/app/features/characters/components/find-character-form/find-character-form.component.ts]: the name must tell what exactly messages are they

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    return EMSG;
+  }
+
+  subs!: Subscription;
+
+  character!: MarvelChar;
+
+  msg: string = EMSG.clear;
```
Comment (Phantomus) [inline src/app/features/characters/components/find-character-form/find-character-form.component.ts]: need to have more descriptive name

File: src/app/features/characters/components/find-character-form/find-character-form.component.html [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  <form [formGroup]="charFinder">
+    <input
+      type="text"
+      name="charName"
+      id="charName"
+      placeholder="Enter name"
+      formControlName="charName"
+      (input)="clearMsg()"
+    >
+    <button (click)="searchCharacter()">FIND</button>
```
Comment (Phantomus) [inline src/app/features/characters/components/find-character-form/find-character-form.component.html]: onSerachCharacter might be better

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+            this.msg = resp ? EMSG.found : EMSG.notFound;
+          }
+        );
+    } else {
+      this.msg = EMSG.emptyInput;
+    }
+  }
+
+  clearMsg() {
```
Comment (Phantomus) [inline src/app/features/characters/components/find-character-form/find-character-form.component.ts]: please provide better descriptive naming

File: src/app/features/characters/pages/character/character.component.ts [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+import { Subscription } from 'rxjs';
+
+@Component({
+  selector: 'character',
+  standalone: true,
+  imports: [],
+  templateUrl: './character.component.html',
+  styleUrl: './character.component.scss'
+})
+export class CharacterComponent {
```
Comment (Phantomus) [inline src/app/features/characters/pages/character/character.component.ts]: CharacterDetails is a better name for this component, doesn't it?

File: src/app/features/characters/pages/character/character.component.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+  subs!: Subscription;
+
+  constructor(
+    private route: ActivatedRoute,
+    private charactersService: CharactersMarvelService) {}
+
+  ngOnInit() {
+    this.id = +(this.route.snapshot.paramMap.get('id') || 0);
+
+    this.subs = this.charactersService.getCharacter(this.id).subscribe(
```
Comment (Phantomus) [inline src/app/features/characters/pages/character/character.component.ts]:
```
this.subs = this.charactersService.getCharacter(this.id)
  .subscribe(res => {
    this.character = res;
  })
  ```

File: src/app/features/characters/characters.component.ts (hunk 4 above)
Comment (Ellyia) [reply src/app/features/characters/characters.component.ts]: I have changed *ngFor and *ngIf syntax on Angular 17 @if and @for

File: src/app/features/characters/characters.component.ts (hunk 5 above)
Comment (Ellyia) [reply src/app/features/characters/characters.component.ts]: changed

File: src/app/features/characters/characters.component.ts (hunk 6 above)
Comment (Ellyia) [reply src/app/features/characters/characters.component.ts]: ok, I've got it

File: src/app/features/characters/components/character-info/character-info.component.html (hunk 7 above)
Comment (Ellyia) [reply src/app/features/characters/components/character-info/character-info.component.html]: I've changed *ngFor on Angular 17 @for

File: src/app/features/characters/components/find-character-form/find-character-form.component.html (hunk 8 above)
Comment (Ellyia) [reply src/app/features/characters/components/find-character-form/find-character-form.component.html]: ok, thank you

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts (hunk 9 above)
Comment (Ellyia) [reply src/app/features/characters/components/find-character-form/find-character-form.component.ts]: changed

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts (hunk 10 above)
Comment (Ellyia) [reply src/app/features/characters/components/find-character-form/find-character-form.component.ts]: changed

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts (hunk 11 above)
Comment (Ellyia) [reply src/app/features/characters/components/find-character-form/find-character-form.component.ts]: changed

File: src/app/features/characters/pages/character/character.component.ts (hunk 14 above)
Comment (Ellyia) [reply src/app/features/characters/pages/character/character.component.ts]: renamed

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts [hunk 16]
```diff
@@ -0,0 +1,63 @@
+import { Component, OnDestroy } from '@angular/core';
+import { FormControl, FormGroup, ReactiveFormsModule } from '@angular/forms';
+import { Router } from '@angular/router';
+import { MarvelChar } from '@characters/models/characters.model';
+import { CharactersMarvelService } from '@characters/services/characters-marvel.service';
+import { Subscription } from 'rxjs';
+import { CharacterSearchStatus } from '@characters/models/characters.enum';
```
Comment (Phantomus) [inline src/app/features/characters/components/find-character-form/find-character-form.component.ts]: CharacterSearchStatusEnum is better

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+    return CharacterSearchStatus;
+  }
+
+  subs!: Subscription;
+
+  character!: MarvelChar;
+
+  searchStatus: CharacterSearchStatus = CharacterSearchStatus.ClearMsg;
+
+  charFinder = new FormGroup({
```
Comment (Phantomus) [inline src/app/features/characters/components/find-character-form/find-character-form.component.ts]: charFinderForm is better, doesn't it?

File: src/app/features/characters/pages/character/character.component.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+import { Subscription } from 'rxjs';
+
+@Component({
+  selector: 'character',
+  standalone: true,
+  imports: [],
+  templateUrl: './character.component.html',
+  styleUrl: './character.component.scss'
+})
+export class CharacterDetailsComponent {
```
Comment (Phantomus) [inline src/app/features/characters/pages/character/character.component.ts]: you also need to rename files according to the class name change

Comment (Ellyia) [reply src/app/features/characters/pages/character/character.component.ts]: done

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts (hunk 17 above)
Comment (Ellyia) [reply src/app/features/characters/components/find-character-form/find-character-form.component.ts]: Who am I to disagree?)

File: src/app/features/characters/components/find-character-form/find-character-form.component.ts [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+  selector: 'find-character-form',
+  standalone: true,
+  imports: [ ReactiveFormsModule ],
+  templateUrl: './find-character-form.component.html',
+  styleUrl: './find-character-form.component.scss'
+})
+export class FindCharacterFormComponent implements OnDestroy {
+  static CharacterSearchStatusEnum = CharacterSearchStatusEnum;
+
+  get charSearchStatus() {
```
Comment (Phantomus) [inline src/app/features/characters/components/find-character-form/find-character-form.component.ts]: why do you need a getter for enum with static data?

Comment (Ellyia) [reply src/app/features/characters/components/find-character-form/find-character-form.component.ts]: I thought, that I need, but I couldn't explain why. I didn't understand how to use CharacterSearchStatusEnum in template. And you have already explained to me, that I don't need it, thank you)

**Review by Phantomus — APPROVED** [review]

**Pull request merged**